site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,685 @@
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Filters []GRPCRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines the backend(s) where matching requests should be
+	// sent.
+
+	// If unspecified or invalid (refers to a non-existent resource or a Service
+	// with no endpoints), the rule performs no forwarding. If there are also no
+	// filters specified that would result in a response being sent, a gRPC `UNAVAILABLE`
+	// status is returned. `UNAVAILABLE` responses must be sent so that the overall
+	// weight is respected; if an invalid backend is requested to have 80% of
+	// requests, then 80% of requests must get a `UNAVAILABLE` instead.
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []GRPCBackendRef `json:"backendRefs,omitempty"`
+}
+
+// GRPCRouteMatch defines the predicate used to match requests to a given
+// action. Multiple match types are ANDed together, i.e. the match will
+// evaluate to true only if all conditions are satisfied.
+//
+// For example, the match below will match a gRPC request only if its service
+// is `foo` AND it contains the `version: v1` header:
+//
+// ```
+// match:
+//   method:
+//     type: Exact
+//     service: "foo"
+//   headers:
+//   - name: "version"
+//     value "v1"
+// ```
+type GRPCRouteMatch struct {
+	// Path specifies a gRPC request service/method matcher. If this field is not
+	// specified, all services and methods will match.
+	//
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"path,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCPathMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method..
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact with method specified)
+	//
+	// Support Custom (Exact with no method specified)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"value,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Method *string `json:"value,omitempty"`
```

Comment:
I find this pretty confusing and we should do a bit of optimization for the easy path.

Based on the defaults, the values are:
```
type: Exact
method: ""
service: ""
```

No matter what we document, this - 'exact match on empty strings' makes little sense to me and I _feel_ it will make little sense to anyone.
If the type is Exact, then the values should contain something.
My experience is that using empty values to mean everything doesn't work well, especially in k8s.

Some options we can explore:
1. Can the code require at least one of method or service?
2. Can we ask the user to provide an explicit `*` as the value of the `method` or `service` to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.
3. If type=Exact, the code should be exacting. Splitting `type` into two types makes the configuration verbose but it also makes it a bit more machine-readable. Let's say the type=regex and then one of them is a regex while the other one is not, how is the code supposed to figure it out?
4. Exact with the method specified is core and no method specified is custom seems arbitrary and confusing. Can we classify both at the same level?

I should note that the above is done with the right technical reasons, I don't doubt. It doesn't feel it is intuitive enough.

Thoughts?

Reply:
> Based on the defaults, the values are:

This is true. Those are the default values. But people generally don't see the default values, they see the default _behavior_. And the default _behavior_ is to match all traffic.

> My experience is that using empty values to mean everything doesn't work well, especially in k8s.

Can you provide some concrete examples?

> Can the code require at least one of method or service?

Considering that supplying no match at all already matches all traffic, I'm totally open to this option. Going ahead and making this change. Anyone objecting, please jump in.

> Can we ask the user to provide an explicit * as the value of the method or service to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.

I'm not the biggest fan of this option. As far as I know, no equivalent syntax exists for HTTPRoute. This could lead to confusion.

> If type=Exact, the code should be exacting. Splitting type into two types makes the configuration verbose but it also makes it a bit more machine-readable.

I don't quite see how this solves the problem you introduced at the beginning of this comment? And as you point out, this would not be as nice a UX.

> Let's say the type=regex and then one of them is a regex while the other one is not, how is the code supposed to figure it out?

Discussed already [here](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826430565).

Specifically, this:

> What's nice here is that exact matches are a subset of regular expressions, so the fact that there's only one type for two values here doesn't actually get in the way of expressing the routing logic you want. 

In practice, this shouldn't be a limitation.

> Exact with the method specified is core and no method specified is custom seems arbitrary and confusing. Can we classify both at the same level?

This actually was mistaken. The combination that presents trouble for implementations is `type: EXACT` with service unspecified and method specified. I've corrected this in the proposal.

I expect that many implementations will add support by translating to HTTPRoutes. In [this chart](https://github.com/kubernetes-sigs/gateway-api/pull/1004#discussion_r826425912) from earlier, you see that this combination is equivalent to either a suffix matcher or a regex matcher of the form `/.+/${METHOD}`. Suffix matchers don't exist for HTTPRoute and Regex matchers have "Custom" support.  This case is called out to have the same support level as an HTTPRoute regex matcher to allow implementations that support HTTPRoute but not regex to also support GRPCRoute at some level.

If you have an alternative to address this problem, I'm open to suggestions.

Reply:
> This is true. Those are the default values. But people generally don't see the default values, they see the default behavior. And the default behavior is to match all traffic.

In that case, the default values _should_ communicate the default behavior well.

> Can you provide some concrete examples?

Empty label selector behavior comes to mind.


> Considering that supplying no match at all already matches all traffic, I'm totally open to this option. Going ahead and making this change. Anyone objecting, please jump in.

Great. Could you create a tracking issue for this? This will require updates to validating webhook; the last time I looked, controller-runtime didn't allow for "at least one" JSON-schema validations. 


> > Can we ask the user to provide an explicit * as the value of the method or service to mean that they would like to match any method or service respectively? IMHO, that makes the configuration much more self-documenting.

> I'm not the biggest fan of this option. As far as I know, no equivalent syntax exists for HTTPRoute.

In HTTPRoute, the default value for path is `/` with the type `PathPrefix`.

Reply:
> Great. Could you create a tracking issue for this? This will require updates to validating webhook; the last time I looked, controller-runtime didn't allow for "at least one" JSON-schema validations.

[Created.](https://github.com/kubernetes-sigs/gateway-api/issues/1098)